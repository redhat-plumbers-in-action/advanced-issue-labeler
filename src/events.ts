//? DOC: https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads

export const events = {
  issue: ['issues.opened' as const],
};
