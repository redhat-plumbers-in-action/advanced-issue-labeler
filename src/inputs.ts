import { IsArray, IsString, MinLength, ValidateIf } from 'class-validator';

import { TInputs } from './types.d';

export class Inputs {
  @ValidateIf(instance => instance._template)
  @IsString()
  @MinLength(1)
  private _template?: string | undefined;

  @IsString()
  @MinLength(1)
  private _section?: string | undefined;

  @ValidateIf(instance => instance._blockList)
  @IsArray()
  @IsString({ each: true })
  @MinLength(1, { each: true })
  private _blockList?: string[] | undefined;

  constructor(inputs: TInputs) {
    this._template = inputs?.template;
    this._section = inputs?.section;
    this._blockList = inputs?.blockList;
  }

  get template() {
    return this._template;
  }

  get section() {
    return this._section;
  }

  get blockList() {
    return this._blockList;
  }
}
