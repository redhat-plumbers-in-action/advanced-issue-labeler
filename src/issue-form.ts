export class IssueForm {
  private _parsed: { [key: string]: string };

  constructor(parsedIssue: {}) {
    this._parsed = parsedIssue;
  }

  private get parsed() {
    return this._parsed;
  }

  isProperty(key: string) {
    return this.parsed.hasOwnProperty(key);
  }

  getProperty(key: string) {
    return this.parsed[key];
  }

  getSafeProperty(key: string) {
    if (!this.isProperty(key)) return undefined;

    return this.getProperty(key);
  }

  listKeywords(key: string, blockList: string[]) {
    const propertySection = this.getSafeProperty(key);

    if (!propertySection) return undefined;

    return propertySection
      .split(', ', 25)
      .filter(label => !this.isCompliant(blockList, label));
  }

  isCompliant(policy: string[], keyword: string) {
    return !!policy.find(rule => keyword === rule);
  }
}
