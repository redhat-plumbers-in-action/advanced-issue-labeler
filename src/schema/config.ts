import { z } from 'zod';

export const configLabelSchema = z.object({
  name: z.string().min(1),
  keys: z.array(z.string().min(1)),
});

export type ConfigLabel = z.infer<typeof configLabelSchema>;

export const configSectionSchema = z.object({
  id: z.array(z.string().min(1)),
  'block-list': z.array(z.string().min(1)).default([]),
  label: z.array(configLabelSchema).default([]),
});

export type ConfigSection = z.infer<typeof configSectionSchema>;

export const configPolicySchema = z.object({
  template: z.array(z.string()).default([]),
  section: z.array(configSectionSchema).default([]),
});

export type ConfigPolicy = z.infer<typeof configPolicySchema>;

export const configSchema = z
  .object({
    policy: z.array(configPolicySchema).min(1).optional().nullable(),
  })
  .optional()
  .nullable();

export type ConfigType = z.infer<typeof configSchema>;
