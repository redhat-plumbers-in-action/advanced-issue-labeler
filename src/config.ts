import { debug } from '@actions/core';
import { Context } from 'probot';
// import { Type } from 'class-transformer';
import {
  ArrayMinSize,
  IsArray,
  IsString,
  MinLength,
  validate,
  ValidateNested,
} from 'class-validator';

import { events } from './events';
import {
  TConfigObject,
  TPolicyItem,
  TSectionItem,
  TLabelItem,
} from './types.d';
import { ValidationFeedback } from './validation-feedback';

export class Config {
  @IsArray()
  @ValidateNested({ each: true })
  @ArrayMinSize(1)
  // @Type(() => SectionItem)
  private _policy: PolicyItem[];

  constructor(config: TConfigObject) {
    this._policy = Array.isArray(config?.policy)
      ? config.policy.map(item => new PolicyItem(item))
      : [];
  }

  get policy() {
    return this._policy;
  }

  getTemplatePolicy(template: string | undefined) {
    // ?: When template name is provided look for section that match the template name
    const templatePolicy = this.policy.find(pItem =>
      pItem.template?.find(tItem => tItem === template)
    );

    // If exact policy for template exists ; return it
    if (templatePolicy) return templatePolicy;

    // ?: When template name is undefined look for section that undefined template field
    debug(`Looking for default policy ...`);
    return this.policy.find(
      pItem =>
        !pItem?.template ||
        (Array.isArray(pItem?.template) && pItem?.template.length === 0)
    );
  }

  static async getConfig(
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    const retrievedConfig = await context.config<TConfigObject>(
      'advanced-issue-labeler.yml'
    );

    if (Config.isConfigEmpty(retrievedConfig)) {
      return null;
    }

    const config = new this(retrievedConfig as Config);

    return config;
  }

  static isConfigEmpty(config: TConfigObject | null | unknown) {
    return config === null;
  }

  static async validate(instance: Config) {
    const validationResult = await validate(instance, {
      whitelist: true,
      forbidNonWhitelisted: true,
    });

    const results = validationResult.map(error => {
      return ValidationFeedback.composeFeedbackObject(error);
    });

    return results;
  }
}

class PolicyItem {
  @IsString({ each: true })
  @MinLength(0, { each: true })
  private _template?: string[];

  @IsArray()
  @ValidateNested({ each: true })
  @ArrayMinSize(1)
  // @Type(() => SectionItem)
  private _section: SectionItem[];

  constructor(item: TPolicyItem) {
    this._template = item?.template ?? [];
    this._section = Array.isArray(item?.section)
      ? item.section.map(sectionItem => new SectionItem(sectionItem))
      : [];
  }

  get template() {
    return this._template;
  }

  get section() {
    return this._section;
  }
}

class SectionItem {
  @IsString({ each: true })
  @MinLength(1, { each: true })
  private _id: string[];

  @IsString({ each: true })
  @MinLength(0, { each: true })
  private _blockList: string[];

  @IsArray()
  @ValidateNested({ each: true })
  @ArrayMinSize(1)
  // @Type(() => SectionItem)
  private _label: TLabelItem[];

  constructor(item: TSectionItem) {
    this._id = item?.id;
    this._blockList = item['block-list'] ?? [];
    this._label = Array.isArray(item?.label)
      ? item.label.map(labelItem => new Label(labelItem))
      : [];
  }

  get id() {
    return this._id;
  }

  get blockList() {
    return this._blockList;
  }

  get label() {
    return this._label;
  }
}

class Label {
  @IsString()
  @MinLength(1)
  private _name: string;

  @IsString({ each: true })
  @MinLength(1, { each: true })
  private _keys: string[];

  constructor(item: TLabelItem) {
    this._name = item?.name;
    this._keys = item?.keys;
  }

  get name() {
    return this._name;
  }

  get keys() {
    return this._keys;
  }
}
