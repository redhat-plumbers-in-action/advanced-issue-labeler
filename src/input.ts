import { z } from 'zod';
import { IssueForm } from './issue-form';
import { inputSchema, InputType, jsonSchema } from './schema/input';

export class Input {
  readonly issueForm: InputType['issueForm'];
  readonly template: InputType['template'];
  readonly section: InputType['section'];
  readonly blockList: InputType['blockList'];

  constructor(input: unknown) {
    const parsedInput = inputSchema.parse(input);

    this.issueForm = parsedInput.issueForm;
    this.template = parsedInput.template;
    this.section = parsedInput.section;
    this.blockList = parsedInput.blockList;
  }
}
