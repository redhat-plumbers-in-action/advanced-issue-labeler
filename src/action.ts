import { getInput, debug, info } from '@actions/core';

import { Config } from './config';
import { IssueForm } from './issue-form';
import { Labeler } from './labeler';
import { CustomOctokit } from './octokit';
import { context } from '@actions/github';

async function action(octokit: CustomOctokit) {
  const issueFormInput = getInput('issue-form', { required: true });

  const labeler = new Labeler(new IssueForm(JSON.parse(issueFormInput)));
  labeler.config = await Config.getConfig(octokit);

  // If no config was provided try inputs
  if (!labeler.isConfig) {
    labeler.setInputs({
      section: getInput('section', { required: true }),
      blockList: getInput('block-list').split('\n', 25),
    });
  }

  // Config requires template as well
  labeler.setInputs({
    template: getInput('template'),
  });

  const validationResult = await Labeler.validate(labeler);

  if (validationResult.length > 0) {
    for (const [index, message] of validationResult.entries()) {
      // TODO: Probably needs some care:
      debug(
        `{ property: ${message.property}, value: ${message.value}, notes: ${message.notes}`
      );
    }
    return;
  }

  const labels = labeler.gatherLabels();

  // Check if there are some labels to be set
  if (!labels || (Array.isArray(labels) && labels?.length < 1)) {
    info('Nothing to do here. CY@');
    return;
  }

  info(`Labels to be set: ${labels}`);

  const response = await octokit.request(
    'POST /repos/{owner}/{repo}/issues/{issue_number}/labels',
    {
      ...context.repo,
      issue_number: context.issue.number,
    }
  );

  debug(`GitHub API response status: [${response.status}]`);
}

export default action;
