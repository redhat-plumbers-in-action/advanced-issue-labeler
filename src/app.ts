import { Context, Probot } from 'probot';
import { getInput, debug } from '@actions/core';

import { events } from './events';
import { Config } from './config';
import { IssueForm } from './issue-form';
import { Labeler } from './labeler';

const app = (probot: Probot) => {
  probot.on(
    events.issue,
    async (context: Context<typeof events.issue[number]>) => {
      const issueFormInput = getInput('issue-form');

      if (!issueFormInput) {
        context.log.error(`Parameter issue-form is required!`);
        return;
      }

      const labeler = new Labeler(new IssueForm(JSON.parse(issueFormInput)));
      labeler.config = await Config.getConfig(context);

      // If no config was provided try inputs
      if (!labeler.isConfig) {
        labeler.inputs = {
          section: getInput('section'),
          blockList: getInput('block-list').split('\n', 25),
        };
      }

      // Config requires template as well
      labeler.inputs = {
        template: getInput('template'),
      };

      const validationResult = await Labeler.validate(labeler);

      if (validationResult.length > 0) {
        for (const [index, message] of validationResult.entries()) {
          // TODO: Probably needs some care:
          probot.log.info(
            `{ property: ${message.property}, value: ${message.value}, notes: ${message.notes}`
          );
          probot.log.warning(`[${index}] ~ Config validation: '${message}'`);
        }
        return;
      }

      const labels = labeler.gatherLabels();

      // Check if there are some labels to be set
      if (!labels || (Array.isArray(labels) && labels?.length < 1)) {
        debug('Nothing to do here. CY@');
        return;
      }

      debug(`Labels to be set: ${labels}`);

      const response = await context.octokit.rest.issues.addLabels(
        context.issue({ labels })
      );

      debug(`GitHub API response: ${response}`);
    }
  );
};

export default app;
