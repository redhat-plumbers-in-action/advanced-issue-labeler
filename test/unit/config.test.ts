import { describe, it, expect, beforeEach, test } from 'vitest';

import { Config } from '../../src/config';
import {
  configContextFixture,
  IConfigTestContext,
} from './fixtures/config-fixture';

describe('Config Object', () => {
  beforeEach<IConfigTestContext>(context => {
    context.configs = configContextFixture.configs;
    context.invalid = configContextFixture.invalid;
  });

  it<IConfigTestContext>('can be instantiated', context => {
    context.configs.map(item => expect(item).toBeDefined());
  });

  test<IConfigTestContext>('get policy()', context => {
    context.configs.map(item => expect(item.policy).toMatchSnapshot());
  });

  it<IConfigTestContext>('is valid', async context => {
    context.configs.map(async item =>
      expect(Config.validate(item)).resolves.toMatchObject([])
    );
  });

  it<IConfigTestContext>('is invalid', async context => {
    context.invalid.map(async item =>
      expect(Config.validate(item)).resolves.toMatchSnapshot()
    );
  });

  it<IConfigTestContext>('can get template policy', context => {
    context.configs.map(async item =>
      expect(item.getTemplatePolicy('issue-template.yml')).toMatchSnapshot()
    );
  });

  test<IConfigTestContext>('is config empty', context => {
    context.configs.map(async item =>
      expect(Config.isConfigEmpty(item)).toEqual(false)
    );

    context.invalid.map(async item =>
      expect(Config.isConfigEmpty(item)).toEqual(false)
    );

    expect(Config.isConfigEmpty(null)).toEqual(true);
  });
});

describe.todo('PolicyItem Object');
describe.todo('SectionItem Object');
describe.todo('Label Object');
