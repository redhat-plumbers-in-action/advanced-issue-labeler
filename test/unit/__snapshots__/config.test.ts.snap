// Vitest Snapshot v1

exports[`Config Object > can get template policy 1`] = `
PolicyItem {
  "_section": [
    SectionItem {
      "_blockList": [],
      "_id": [
        "type",
      ],
      "_label": [
        Label {
          "_keys": [
            "Bug Report",
          ],
          "_name": "bug 游냍",
        },
        Label {
          "_keys": [
            "Feature Request",
          ],
          "_name": "RFE 游꾸",
        },
      ],
    },
  ],
  "_template": [],
}
`;

exports[`Config Object > can get template policy 2`] = `
PolicyItem {
  "_section": [
    SectionItem {
      "_blockList": [
        "Other",
      ],
      "_id": [
        "type",
      ],
      "_label": [
        Label {
          "_keys": [
            "Bug Report",
          ],
          "_name": "bug 游냍",
        },
        Label {
          "_keys": [
            "Feature Request",
          ],
          "_name": "RFE 游꾸",
        },
      ],
    },
  ],
  "_template": [
    "issue-template.yml",
  ],
}
`;

exports[`Config Object > can get template policy 3`] = `
PolicyItem {
  "_section": [
    SectionItem {
      "_blockList": [
        "Other",
      ],
      "_id": [
        "type",
      ],
      "_label": [
        Label {
          "_keys": [
            "Bug Report",
          ],
          "_name": "bug 游냍",
        },
        Label {
          "_keys": [
            "Feature Request",
          ],
          "_name": "RFE 游꾸",
        },
      ],
    },
  ],
  "_template": [
    "issue-template.yml",
  ],
}
`;

exports[`Config Object > can get template policy 4`] = `
PolicyItem {
  "_section": [
    SectionItem {
      "_blockList": [
        "Other",
      ],
      "_id": [
        "type",
      ],
      "_label": [
        Label {
          "_keys": [
            "Bug Report",
          ],
          "_name": "bug 游냍",
        },
        Label {
          "_keys": [
            "Feature Request",
          ],
          "_name": "RFE 游꾸",
        },
      ],
    },
  ],
  "_template": [
    "issue-template.yml",
    "template.yml",
  ],
}
`;

exports[`Config Object > get policy() 1`] = `
[
  PolicyItem {
    "_section": [
      SectionItem {
        "_blockList": [],
        "_id": [
          "type",
        ],
        "_label": [
          Label {
            "_keys": [
              "Bug Report",
            ],
            "_name": "bug 游냍",
          },
          Label {
            "_keys": [
              "Feature Request",
            ],
            "_name": "RFE 游꾸",
          },
        ],
      },
    ],
    "_template": [],
  },
]
`;

exports[`Config Object > get policy() 2`] = `
[
  PolicyItem {
    "_section": [
      SectionItem {
        "_blockList": [
          "Other",
        ],
        "_id": [
          "type",
        ],
        "_label": [
          Label {
            "_keys": [
              "Bug Report",
            ],
            "_name": "bug 游냍",
          },
          Label {
            "_keys": [
              "Feature Request",
            ],
            "_name": "RFE 游꾸",
          },
        ],
      },
    ],
    "_template": [
      "issue-template.yml",
    ],
  },
]
`;

exports[`Config Object > get policy() 3`] = `
[
  PolicyItem {
    "_section": [
      SectionItem {
        "_blockList": [
          "Other",
        ],
        "_id": [
          "type",
        ],
        "_label": [
          Label {
            "_keys": [
              "Bug Report",
            ],
            "_name": "bug 游냍",
          },
          Label {
            "_keys": [
              "Feature Request",
            ],
            "_name": "RFE 游꾸",
          },
        ],
      },
    ],
    "_template": [
      "issue-template.yml",
    ],
  },
  PolicyItem {
    "_section": [
      SectionItem {
        "_blockList": [
          "other",
        ],
        "_id": [
          "id",
        ],
        "_label": [
          Label {
            "_keys": [
              "label",
            ],
            "_name": "label",
          },
        ],
      },
    ],
    "_template": [
      "template1.yml",
      "template1.yml",
    ],
  },
]
`;

exports[`Config Object > get policy() 4`] = `
[
  PolicyItem {
    "_section": [
      SectionItem {
        "_blockList": [
          "Other",
        ],
        "_id": [
          "type",
        ],
        "_label": [
          Label {
            "_keys": [
              "Bug Report",
            ],
            "_name": "bug 游냍",
          },
          Label {
            "_keys": [
              "Feature Request",
            ],
            "_name": "RFE 游꾸",
          },
        ],
      },
    ],
    "_template": [
      "issue-template.yml",
      "template.yml",
    ],
  },
  PolicyItem {
    "_section": [
      SectionItem {
        "_blockList": [
          "other",
        ],
        "_id": [
          "id",
        ],
        "_label": [
          Label {
            "_keys": [
              "label",
            ],
            "_name": "label",
          },
        ],
      },
    ],
    "_template": [
      "template1.yml",
      "template1.yml",
    ],
  },
]
`;

exports[`Config Object > is invalid 1`] = `
[
  {
    "notes": {
      "arrayMinSize": "_policy must contain at least 1 elements",
    },
    "property": "_policy",
    "value": [],
  },
]
`;

exports[`Config Object > is invalid 2`] = `
[
  {
    "notes": {
      "arrayMinSize": "_policy must contain at least 1 elements",
    },
    "property": "_policy",
    "value": [],
  },
]
`;

exports[`Config Object > is invalid 3`] = `
[
  {
    "notes": {
      "arrayMinSize": "_policy must contain at least 1 elements",
    },
    "property": "_policy",
    "value": [],
  },
]
`;

exports[`Config Object > is invalid 4`] = `
[
  {
    "notes": {
      "arrayMinSize": "_label must contain at least 1 elements",
    },
    "property": "_policy.0._section.0._label",
    "value": [],
  },
]
`;

exports[`Config Object > is invalid 5`] = `
[
  {
    "notes": {
      "arrayMinSize": "_section must contain at least 1 elements",
    },
    "property": "_policy.0._section",
    "value": [],
  },
]
`;

exports[`Config Object > is invalid 6`] = `
[
  {
    "notes": {
      "isString": "each value in _keys must be a string",
      "minLength": "each value in _keys must be longer than or equal to 1 characters",
    },
    "property": "_policy.0._section.0._label.1._keys",
    "value": undefined,
  },
]
`;
